rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php

```diff
@@ -0,0 +1,162 @@
+use PhpParser\Node\Scalar;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Stmt\Return_;
+use PHPStan\Type\ConstantType;
+use PHPStan\Type\GeneralizePrecision;
+use PHPStan\Type\Type;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\ValueObject\PhpVersionFeature;
+use Rector\PHPStanStaticTypeMapper\Enum\TypeKind;
+use Rector\VersionBonding\Contract\MinPhpVersionInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\TypeDeclaration\Rector\Class_\ReturnTypeFromStrictTernaryRector\ReturnTypeFromStrictTernaryRectorTest
+ */
+final class ReturnTypeFromStrictTernaryRector extends AbstractRector implements MinPhpVersionInterface
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add method return type based on strict ternary values', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function getValue($number)
+    {
+        return $number ? 100 : 500;
+    }
+}
+CODE_SAMPLE
+
+                ,
+                <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function getValue($number): int
+    {
+        return $number ? 100 : 500;
+    }
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $hasChanged = false;
+
+        foreach ($node->getMethods() as $classMethod) {
+            if ($classMethod->returnType instanceof Node) {
+                continue;
+            }
+
+            $onlyStmt = $classMethod->stmts[0] ?? null;
+            if (! $onlyStmt instanceof Return_) {
+                continue;
+            }
+
+            if (! $onlyStmt->expr instanceof Ternary) {
+                continue;
+            }
+
+            $ternary = $onlyStmt->expr;
+
+            // has scalar in if/else of ternary
+            if (! $this->isAlwaysScalarExpr($ternary->if)) {
+                continue;
+            }
+
+            if (! $this->isAlwaysScalarExpr($ternary->else)) {
+                continue;
+            }
+
+            /** @var Node\Expr $if */
+            $if = $ternary->if;
+
+            /** @var Node\Expr $else */
+            $else = $ternary->else;
+
+            $ifType = $this->staticTypeMapper->mapPhpParserNodePHPStanType($if);
+            $elseType = $this->staticTypeMapper->mapPhpParserNodePHPStanType($else);
+
+            if (! $this->areTypesEqual($ifType, $elseType)) {
+                continue;
+            }
+
+            $returnTypeNode = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode($ifType, TypeKind::RETURN);
+
+            if (! $returnTypeNode instanceof Node) {
+                continue;
+            }
+
```

Comment:
parent classmethod override guard is needed here

https://github.com/rectorphp/rector-src/blob/b98de3edc21312c2e5483b6a120ccb04642f5346/rules/TypeDeclaration/Rector/ClassMethod/AddReturnTypeDeclarationRector.php#L145

Reply:
Skip classmethod also exist https://github.com/rectorphp/rector-src/blob/b98de3edc21312c2e5483b6a120ccb04642f5346/rules/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromReturnDirectArrayRector.php#L79